### PR TITLE
[CDAP-17709] Adds support for PROXY authentication mode in UI

### PR DESCRIPTION
**Pull request by ajainarayanan**
JIRA - https://cdap.atlassian.net/browse/CDAP-17709

**Context**
- CDAP UI now supports both MANAGED and PROXY authentication modes
- MANAGED authentication mode is the regular type of authentication that we used to support
- PROXY mode is when the authentication happens behind a inverting proxy with client being unaware of it.

**Auth Flow for PROXY mode**
- User navigates to CDAP behind an inverting proxy.
- Inverting proxy checks for user authentication and redirects to oAuth provider if not authenticated
- If authenticated, adds auth token and user id and passes on the request to nodejs server
- Node server then passes on the UI assets for initial page load
- Subsequent HTTP requests follow the same path and nodejs server passes on the auth token and user id in the headers to the backend.

**Websockets in PROXY mode**
- Since websockets are not designed to handle authorization or authentication we need to handle it before the connection is established
- When browser initiates the initial handshake for websocket it sends a plain HTTP request which then get upgraded to websocket.
- During this upgrade call the nodejs registers the auth token and user id and on each subsequent ws connection adds them to the connection object.
- This helps us in adding them to headers in the request we make to the backend.

**Session token**
- As part of this change we are also removing the auth token to be used while generating the session token
- The reason being once a session token is tied to the auth token we then have to know to refresh the session when auth token expires
- Since session token is only to verify the client session it can be generated without auth token.

### CODE REVIEW

File: cdap-ui/app/services/data/my-cdap-datasource.js [hunk 1]
```diff
@@ -29,7 +29,11 @@ angular.module(PKG.name + '.services')
 
       // FIXME: There is a circular dependency and that is why
       // myAuth.isAuthenticated is not used. There should be a better way to do this.
-      if (window.CDAP_CONFIG.securityEnabled && $rootScope.currentUser && $rootScope.currentUser.token) {
+      if (
```
Comment (njbriggs) [inline cdap-ui/app/services/data/my-cdap-datasource.js]: Any chance you can address the `FIXME` above?

Comment (ajainarayanan) [reply cdap-ui/app/services/data/my-cdap-datasource.js]: Fair point but as of now no. Ideally I don't want to spend time on this when we know this is going to go away. I would need to spend sometime in investigating why the cyclic dependency happens and clean it up and at this point I don't want to introduce regressions 😭

File: cdap-ui/server/aggregator.js [hunk 2]
```diff
@@ -211,6 +211,11 @@ function stripResource(key, value) {
 function emitResponse(resource, error, response, body) {
   var timeDiff = Date.now() - resource.startTs;
 
+  if (response && (response.statusCode === 401 || response.statusCode === 403)) {
```
Comment (njbriggs) [inline cdap-ui/server/aggregator.js]: Should we do this on a `403`? `403` should mean the client is authenticated, but attempted to access a resource they don't have permissions on.

File: cdap-ui/server/token.mjs [hunk 3]
```diff
@@ -117,7 +117,7 @@ function getSecretFromCDAPConfig(cdapConfig, logger) {
  * 1. Get the following from cdap config (cdap-site + cdap-security)
  *    - Secret key, - secret
  *    - environment specific ID, - instancename
- * 2. Create a hash of (creation time + authToken + instance name)
+ * 2. Create a hash of (creation time + instance name)
```
Comment (njbriggs) [inline cdap-ui/server/token.mjs:120]: A collision between different users could occur if they hit this function at the same time. If there's a collision, will that weaken the overall token?

File: cdap-ui/app/services/data/my-cdap-datasource.js (hunk 1 above)
Comment (njbriggs) [reply cdap-ui/app/services/data/my-cdap-datasource.js]: That's fine. I was really wondering if the new changes made it moot.

Comment (ajainarayanan) [reply cdap-ui/app/services/data/my-cdap-datasource.js]: No. The dependency still exists.

File: cdap-ui/server/aggregator.js (hunk 2 above)
Comment (ajainarayanan) [reply cdap-ui/server/aggregator.js]: Fair point. I will check only for 401 status 👍

File: cdap-ui/server/token.mjs (hunk 3 above)
Comment (ajainarayanan) [reply cdap-ui/server/token.mjs:120]: Session token here is used only to validate the client connection. Anybody can get a sessiontoken from nodejs server from the `/sessionToken` endpoint. It is ok if there is a collision as it only validates the client and not share any sensitive data.

File: cdap-ui/app/services/data/my-cdap-datasource.js [hunk 4]
```diff
@@ -29,9 +29,9 @@ angular.module(PKG.name + '.services')
 
       // FIXME: There is a circular dependency and that is why
       // myAuth.isAuthenticated is not used. There should be a better way to do this.
-      if (window.CDAP_CONFIG.securityEnabled && $rootScope.currentUser && $rootScope.currentUser.token) {
+      if (window.CDAP_CONFIG.securituEnabled && $cookies.get('CDAP_Auth_Token')) {
```
Comment (elfenheart) [inline cdap-ui/app/services/data/my-cdap-datasource.js]: typo: `security`

**Review by njbriggs — APPROVED** [review]

**Review by elfenheart — APPROVED** [review]

**Pull request merged**